File: python/LinkedList/LinkedList.py
from Node import Node

class LinkedList:

	# Refatorar isso aqui, 
 	# porque uma linked list pode ser iniciada vazia(?)
    def __init__(self, value):
        self._head = Node(value)
        self._tail = self._head
        self._length = 1

    @property
    def head(self):
        return self._head
    
    @property
    def tail(self):
        return self._tail


    @property
    def length(self):
        return self._length


    def clear(self):
        self._head = None
        self._tail = None
        self._length = 0


    def insertAtBeginning(self, value):
        if self._length==0:
            self._head = Node(value)
            self._tail = self.head
            self._length+=1
        else:
            aux = Node(value)
            aux.next = self._head
            self._head = aux
            self._length +=1

    def insertAtEnd(self,value):
        if self._length==0:
            self._head = Node(value)
            self._tail = self._head
            self._length+=1
        else:
            aux = Node(value)
            self._tail.next = aux 
            self._tail = aux
            self._length+=1

    def insertAtIndex(self, value, n):
        cur = self._head
        aux = Node(value)
        for i in range(n-2):
            cur = cur.next
        aux.next = cur.next
        cur.next = aux
        self._length+=1

    def deleteNode(self, value):
        curr = self._head
        if(curr.value==value):
            if(self._length==1):
                self._head = None
                self._length-=1
                self._tail = None
            else:
                self._head = self._head.next
                self._length-=1
        else:
            while(curr.next!=None):
                if(curr.next.value==value):
                    curr.next = curr.next.next
                    self._length-=1
                    break
                else:
                    curr=curr.next
            print('valor nao encontrado!')  

    def deleteAtBeginning(self):
        self._head = self._head.next
        self._length-=1

    def deleteAtEnd(self):
        cur = self._head
        for i in range(self._length-2):
            cur=cur.next
        cur.next=None
        self._tail= cur
        self._length-=1
        
    
    def searchValue(self, value):
        cur=self._head
        while(cur.next!=None):
            if(cur.value==value):
                return(cur.value)
            else:
                cur=cur.next
            
        if cur.value==value:
            return(cur.value)
        else:
            return('Valor nao encontrado')


    def isEmpty(self):
        if self._length ==0:
            return True
        else:
            return False

    
    def printList(self):
        if self._length ==0:
            print('lista vazia!')
        else:
            lista = ''
            curr = self._head
            while(curr.next!=None):
                lista += str(curr.value)+' '
                curr = curr.next
            lista+=str(curr.value)
            print(lista)
    
    def swapNodes(self, val1, val2):
        
        if val1 == val2:
            print("Elements are the same - no swap needed")
            return
 
        node1_prev = None
        node2_prev = None
        node1 = self._head
        node2 = self._head
        
        while node1 is not None:
            if node1.value == val1:
                break
            node1_prev = node1
            node1 = node1.next
            
        while node2 is not None:
            if node2.value==val2:
                break
            node2_prev = node2
            node2 = node2.next
        
        if(node1 is None or node 2 is None):
            print("Swap not possible - one of more elemnt is not in the list")
            return
            
        if node1_prev == None:
            self._head=node2
        else:
            node1_prev.next = node2
            
        if node2_prev == None:
            self._head=node1
        else:
            node2_prev.next = node1
                  
        temp = node1.next
        node1.next = node2.next
        node2.next = temp
    
            

File: javascript/LinkedList/LinkedList.js
const Node =require('./Node.js');

// Operações que vamos implantar nas linked lists:

// head - Returns the head node of the LinkedList. ok
// tail - Returns the tail node of the LinkedList. ok
// length - Returns the length (number of nodes) in the LinkedList. ok 
// clear - Clears the LinkedList by removing all nodes. ok
// insertAtBeginning - Inserts a new node with the given value at the beginning of the LinkedList. ok
// insertAtEnd - Inserts a new node with the given value at the end of the LinkedList. ok
// insertAtIndex - Inserts a new node with the given value at the specified index in the LinkedList. ok
// deleteNode - Deletes the node with the given value from the LinkedList. ok
// deleteAtBeginning - Deletes the first node from the LinkedList. ok Falta testar
// deleteAtEnd - Deletes the last node from the LinkedList. ok Falta testar
// searchValue - Searches for the given value in the LinkedList and returns it if found. ok Falta testar
// isEmpty - Checks if the LinkedList is empty and returns a boolean value. ok
// printList - Prints the values of all nodes in the LinkedList. ok


class LinkedList{

    constructor(value){
        this._head = new Node(value);
        this._tail = this._head;
        this._length = 1;
    }

    get head(){
        return this._head;
    }

    get tail(){
        return this._tail;
    }

    get length(){
        return this._length;
    }

    clear(){
        this._head = null;
        this._tail = null;
        this._length = 0;
    }

    insertAtBeginning(value){
        if(this._length==0){
            const newNode = new Node(value);
            this._head = newNode;
            this._tail = this._head;
            this._length+=1;    
        } else {
            const aux = new Node(value);
            aux._next = this._head;
            this._head = aux;
            this._length+=1;
        }
    }

    insertAtEnd(value){
        if(this._length==0){
            const newNode = new Node(value);
            this._head = newNode;
            this._tail = this._head;
            this._length += 1;     
        } else {
            const aux = new Node(value);
            this._tail._next=aux;
            this._tail = aux;
            this._length+=1;
        }
    }

    insertAtIndex(value, position){
        if(this._length+1<position){
            console.log('Lista de tamanho menor do que a posicao!')
        } else {

            const newNode = new Node(value);
            let curr = this._head;
            for(let i=0;i<position-2;i++){
                curr = curr._next;
            }
            newNode._next = curr._next;
            curr._next = newNode;
        }

    }

    deleteNode(value){
        let curr = this._head;
        if(curr.value ===value){
            if(this._length===1){
                this._head = null;
                this._tail =null;
                this._length--;
            } else {
                this._head= this._head._next
            }
        } else {
            while(curr._next!==null){
                if(curr._next._value===value){
                    curr._next = curr._next._next
                    break;
                } else {
                    curr=curr._next;
                }
            console.log('Valor não encontrado!')
            }
        }
    }


    printList(){
        let lista = '';
        let aux = this._head;
        if(this._length===0){
            console.log('lista vazia')
        } else {
            while(aux._next!=null){
                lista+=aux._value+' ';
                aux = aux._next; 
            }
            lista+=aux._value;
            console.log(lista)
        }
        
    }

    deleteAtBeginning(){
        if(this._length===0){
            console.log('Lista vazia!')
        } else if(this._length==1){
            this._head=null;
            this._tail=null;
            this._length=0;
        } else {
            this._head=this._head.next;
            this._length--;
        }
    }
    

    deleteAtEnd(){
        if(this._length===0){
            console.log('Lista vazia!')
        } else if(this._length==1){
            this._head=null;
            this._tail=null;
            this._length=0;
        } else {
            let curr = this._head
            for(let i = 1; i<this._length-1< i++){
                curr=curr.next;
            }
            this._tail = curr;
            curr._next= null;
            this._length--;
        }

    }

    searchValue(value){
        let curr = this._head;
        if(this._length===0){
            console.log('empty list!');
        } else {
            while(curr._value!==value){
                if(curr._next!==null){
                    console.log('value not found!');
                } else {
                    curr=curr._next;
                }
            }
        }
    }

    isEmpty(){
        if(this._length===0){
            return true;
        }
    }
}


module.exports = LinkedList;